### PR TITLE
remove notification key from FCM payload, add content-available: true & use higher priority

### PR DESCRIPTION
**Pull request by alexmercerind**
In Flutter version of the application, we show the notifications programmatically because custom image & buttons are added into the notification (controlled entirely inside Dart/Flutter code).

Having `notification` in the FCM payload causes the Firebase to show notification automatically (with no control over it). This in turn, causes the application to show same notification twice (one from the Firebase without any buttons or image & the other one sent by Flutter).

Now the `data` inside the payload is consumed to show the notification & other details present in it.

Reference: https://github.com/firebase/flutterfire/discussions/6113.

It'll be likely better to have this in a separate branch.

### CODE REVIEW

**Comment by alexmercerind** [comment]
Updated PR. Notifications issued by our client app have higher priority (in general).

We need the application to be "awaken" from background for sending notifications from Flutter, because we are only consuming the data from `payload` (and not `notification`). `contentAvailable: true` causes inactive client app on iOS to be awoken. On Android, `data` messages wake the app by default.

`priority: 'high'` makes the `data` to be received immediately on the client app. By default, `data`-only payload shows with `priority: 'normal'` (causes some unspecified delay, battery-consumption etc). Since, `notification` key is now removed from the payload, it is important to set the `priority` to `high` explicitly.

A discussion regarding this can be seen at:
https://github.com/firebase/flutterfire/issues/8277#issuecomment-1152535863
https://firebase.google.com/docs/cloud-messaging/http-server-ref

<hr></hr>

Let's hope it works.

File: index.js [hunk 1]
```diff
@@ -24,10 +24,6 @@ const router = express.Router();
 async function handleMotionEvent(clientTokens, {server_uuid, event_datetime, device_id, device_name, dvr_name}) {
 
     return admin.messaging().sendToDevice(Object.values(clientTokens), {
-        notification: {
```
Comment (tugrul) [inline index.js:27]: removing this section can broke actual mobile app

File: index.js [hunk 2]
```diff
@@ -37,40 +33,41 @@ async function handleMotionEvent(clientTokens, {server_uuid, event_datetime, dev
...
+        contentAvailable: true,
+        priority: 'high',
     });
 
 }
 
 async function handleDeviceState(clientTokens, {server_uuid, event_datetime, device_id, device_name, state, dvr_name}) {
 
     return admin.messaging().sendToDevice(Object.values(clientTokens), {
-        notification: {
```
Comment (tugrul) [inline index.js:47]: removing this section can broke actual mobile app

File: index.js [hunk 3]
```diff
@@ -37,40 +33,41 @@ async function handleMotionEvent(clientTokens, {server_uuid, event_datetime, dev
...
+        contentAvailable: true,
+        priority: 'high',
     });
 
 }
 
 async function handleSolo(clientTokens, {server_uuid, event_datetime, state, dvr_name}) {
 
     return admin.messaging().sendToDevice(Object.values(clientTokens), {
-        notification: {
```
Comment (tugrul) [inline index.js:65]: removing this section can broke actual mobile app

**Review by tugrul — COMMENTED** [review]
Did you try to use background message handler on the client side?

**Comment by alexmercerind** [comment]
Hey @tugrul, may you please review the pull-request again.

I exposed `disable_payload_notification` in the /store-token POST body.

If this is passed `true` ([see here](https://github.com/bluecherrydvr/unity/blob/fce2aad3213298f70e91eb549a71699826e5c6e4/lib/api/api.dart#L220)), then notification won't be included in the FCM payload object.
I hope everything is correct & won't break the React Native version of the application now.

Thanks!

**Comment by alexmercerind** [comment]
> Did you try to use background message handler on the client side?

Yeah.

File: index.js [hunk 4]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
-            deviceName: device_name,
-            dvrName: dvr_name
-
+async function handleMotionEvent(clientTokens, clientDisablePayloadNotifications, {server_uuid, event_datetime, device_id, device_name, dvr_name}) {
+    const clientTokensWithoutNotification = [], clientTokensWithNotification = [];
+    for (const key in clientTokens) {
+        if (clientDisablePayloadNotifications[key] == true) {
+            clientTokensWithoutNotification.push(clientTokens[key]);
+        } else {
+            clientTokensWithNotification.push(clientTokens[key]);
```
Comment (tugrul) [inline index.js]: make this separation before method call. this cause bloated code.

File: index.js [hunk 5]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
-            deviceId: String(device_id),
-            deviceName: device_name,
-            dvrName: dvr_name
+async function handleDeviceState(clientTokens, clientDisablePayloadNotifications, {server_uuid, event_datetime, device_id, device_name, state, dvr_name}) {
+    const clientTokensWithoutNotification = [], clientTokensWithNotification = [];
+    for (const key in clientTokens) {
+        if (clientDisablePayloadNotifications[key] == true) {
+            clientTokensWithoutNotification.push(clientTokens[key]);
+        } else {
+            clientTokensWithNotification.push(clientTokens[key]);
```
Comment (tugrul) [inline index.js]: make this separation before method call. this cause bloated code.

File: index.js [hunk 6]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
-            eventType: 'solo',
-            eventDateTime: String(event_datetime),
-            dvrName: dvr_name
+async function handleSolo(clientTokens, clientDisablePayloadNotifications, {server_uuid, event_datetime, state, dvr_name}) {
+    const clientTokensWithoutNotification = [], clientTokensWithNotification = [];
+    for (const key in clientTokens) {
+        if (clientDisablePayloadNotifications[key] == true) {
+            clientTokensWithoutNotification.push(clientTokens[key]);
+        } else {
+            clientTokensWithNotification.push(clientTokens[key]);
```
Comment (tugrul) [inline index.js]: make this separation before method call. this cause bloated code.

File: index.js [hunk 7]
```diff
@@ -142,6 +221,7 @@ router.post('/hook', wrap(async (req, res) => {
     }
 
     const clientTokens = await redisClient.hgetall('server_token:' + server_uuid);
+    const clientDisablePayloadNotifications = await redisClient.hgetall('server_disable_payload_notification:' + server_uuid);
```
Comment (tugrul) [inline index.js:206]: split clientTokens two groups of array by clientDisablePayloadNotifications

File: index.js [hunk 8]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
+        } else {
+            clientTokensWithNotification.push(clientTokens[key]);
         }
-    }, {
-        contentAvailable: true,
-        priority: 'high',
-    });
+    }
+    return Promise.all([
+        admin.messaging().sendToDevice(clientTokensWithoutNotification, {
```
Comment (tugrul) [inline index.js]: make new method for this like `handleMotionEventWithoutNotification`

File: index.js [hunk 9]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
```
Comment (tugrul) [inline index.js]: make new method for this like `handleDeviceStateWithoutNotification`

File: index.js [hunk 10]
```diff
@@ -21,54 +21,129 @@ const router = express.Router();
...
```
Comment (tugrul) [inline index.js]: make new method for this like `handleSoloWithoutNotification`

File: index.js [hunk 11]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
 
     switch (event_name) {
         case 'motion_event':
-            await handleMotionEvent(clientTokens, req.body);
```
Comment (tugrul) [inline index.js:156]: don't remove `await handleMotionEvent(clientTokensWithNotification, req.body);` call it with notification array.

File: index.js [hunk 12]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
...
+            await handleMotionEvent(clientTokens, clientDisablePayloadNotifications, req.body);
```
Comment (tugrul) [inline index.js]: call here `await handleMotionEventWithoutNotification(clientTokensWithoutNotification, req.body);`

File: index.js [hunk 13]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
...
             break;
         case 'device_state':
-            await handleDeviceState(clientTokens, req.body);
```
Comment (tugrul) [inline index.js:159]: don't remove await `await handleDeviceState(clientTokensWithNotification, req.body);` call it with notification array.

File: index.js [hunk 14]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
...
+            await handleDeviceState(clientTokens, clientDisablePayloadNotifications, req.body);
```
Comment (tugrul) [inline index.js]: call here `await handleDeviceStateWithoutNotification(clientTokensWithoutNotification, req.body);`

File: index.js [hunk 15]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
...
             break;
         case 'solo':
-            await handleSolo(clientTokens, req.body);
```
Comment (tugrul) [inline index.js:162]: don't remove `await handleDeviceState(clientTokensWithNotification, req.body);` call it with notification array.

File: index.js [hunk 16]
```diff
@@ -150,13 +230,13 @@ router.post('/hook', wrap(async (req, res) => {
...
+            await handleSolo(clientTokens, clientDisablePayloadNotifications, req.body);
```
Comment (tugrul) [inline index.js]: call here `await handleSoloWithoutNotification(clientTokensWithoutNotification, req.body)`;

**Review by tugrul — CHANGES_REQUESTED** [review]

**Comment by alexmercerind** [comment]
Hi @tugrul!
I've addressed all the requested changes.
Could you please have a look / review again.

Thanks!

**Comment by tugrul** [comment]
Thank you @alexmercerind. Looks good.

Can you combine all four commits into one commit? You can use interactive rebase functionality of git for this. https://www.youtube.com/watch?v=JkpYvXdbnfQ

**Comment by alexmercerind** [comment]
> Thank you @alexmercerind. Looks good.
> 
> Can you combine all four commits into one commit? You can use interactive rebase functionality of git for this. 

Thanks! There's only one commit now.

**Review by tugrul — APPROVED** [review]

**Pull request merged**